src/librustc_mir/build/block.rs

```diff
@@ -79,4 +80,31 @@ impl<'a,'tcx> Builder<'a,'tcx> {
             block.unit()
         })
     }
+
+    // Helper method for generating MIR inside a conditional block.
+    pub fn with_cond<F>(&mut self, block: BasicBlock, span: Span,
```

Comment:
[1]: being the prime suspect…

Instead of using the odd closure approach, I would just return `(BasicBlock, BasicBlock)`.

Compare:

```diff
-                    block = this.with_cond(
-                        block, expr_span, Operand::Consume(is_min), |this, block| {
-                            this.panic(block, "attempted to negate with overflow", expr_span);
-                            block
-                        }); 
+                    (block, false_block) = this.cfg.cond(expr_span, Operand::Consume(is_min));
+                    this.panic(false_block, "attempted to negate with overflow", expr_span);
```

Note that this not only becomes much more concise, but also doesn’t need any of these hacks like I nitd below.